File: src/player/model/PlaybackEngine.java
package player.model;

import audio.*;
import audio.javafx.JavaFXAudioEngine;
import audio.javasound.JavaSoundEngine;
import distributed.DFile;
import distributed.Peer;
import player.model.data.PlaybackStatus;
import player.model.data.PlayerTarget;
import player.model.data.Speaker;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Controls the isLocal audio engine.
 * An instance of PlaybackEngine observe a PlayerStatus.
 * Whenever an audio file should be played back on a isLocal device, the PlaybackEngine loads the file and plays it back.
 * It also adjusts position and gain to match the desired status.
 *
 * When a playback event occurs like an error or end-of-file, the PlaybackEngine updates the status.
 */
public class PlaybackEngine {
	private PlayerTarget target;
	private PlaybackStatus info;
	private AudioEngine audio;
	private final List<String> supportedTypes;

	private Optional<DFile> currentFile = Optional.empty(); // locally playing media
	private Player player;
	private long lastPositionUpdate;

	private double gain;
	private boolean mute;
	private boolean endOfMediaReached;
	private String errorMessage;

	private List<Speaker> speakers;


	private PlaybackEngine(PlayerTarget target, PlaybackStatus status, AudioEngine audioEngine) throws AudioEngineException {
		this.target = target;
		this.info = status;

		audio = audioEngine;
		supportedTypes = new ArrayList<>(audio.getSupportedMediaTypes().stream().map(t -> t.getFileExtension()).collect(Collectors.toList()));

		// Publish audio devices
		speakers = audio.getDevices().stream()
				.map(dev -> new Speaker(Peer.getLocal(), dev.getID(), dev.getName(), dev.getMinGain(), dev.getMaxGain(), dev.isDefault()))
				.collect(Collectors.toList());

		// Set device if not present
		if(!target.getTargetDevice().isPresent()) {
			target.setTargetDevice(Optional.of(speakers.get(0)));
		}

		target.addDataChangeListener(e -> targetChanged());
	}


	public List<Speaker> getSpeakers() {
		return speakers;
	}


	public static PlaybackEngine initializeAudioEngine(PlayerTarget target, PlaybackStatus status, String engineName) throws AudioEngineException {
		AudioEngine engine;
		if(engineName == null) engine = new JavaSoundEngine();
		else if(engineName.equals("java")) engine = new JavaSoundEngine();
		else if(engineName.equals("javafx")) engine = new JavaFXAudioEngine();
		else throw new AudioEngineException("No audio engine registered for name " + engineName);
		return new PlaybackEngine(target, status, engine);
	}


	private void targetChanged() {
		if(isTarget()) {
			loadFile();
			gain = target.getTargetGain();
			mute = target.isTargetMute();
			endOfMediaReached = false;
			if(player != null) adjustPlayer();
			publishInfo();
		}
		else if(!audio.getPlayers().isEmpty()){
			audio.getPlayers().forEach(Player::dispose);
		}
	}


	private void loadFile() {
		if(!currentFile.equals(target.getTargetMedia())) {
			// Change file

			if(player != null) player.dispose();
			player = null;

			publishInfo();

			currentFile = target.getTargetMedia();
			if(!currentFile.isPresent()) return;

            MediaFile file = new DMediaFile(currentFile.get());
            player = audio.newPlayer(file);
            try {
                player.prepare();
                player.activate(audio.getDefaultDevice());
                player.setMute(mute);
                player.setGain(gain);
                player.addEndOfMediaListener(e -> {endOfMediaReached = true; publishInfo(); });
                if(player.getDuration() < 0) {
                    new Thread(() -> {
                        try {
                            player.waitForDurationProperty();
                        } catch (IllegalStateException e1) {
                            e1.printStackTrace();
                            return;
                        } catch (InterruptedException e1) {
                            return;
                        }
                        publishInfo();
                    }).start();
                }
                errorMessage = null;
            } catch(Exception exc) {
                player = null;
                exc.printStackTrace();
                errorMessage = exc.getMessage();
            }
		}
	}


	private void adjustPlayer() {
		AudioDevice localDevice = findLocalDevice(target.getTargetDevice()).get();
		if(mute != player.isMute()) {
			player.setMute(mute);
		}
		if(gain != player.getGain()) {
			player.setGain(gain);
		}
		if(target.wasTargetPositionSetAfter(lastPositionUpdate)) {
			lastPositionUpdate = target.getPositionUpdateTime();
			try {
				player.setPositionBlocking(target.getTargetPosition().getAsDouble(), 1.0);
			} catch (InterruptedException e1) {
				// TODO Auto-generated catch block
				e1.printStackTrace();
			} catch (IOException e1) {
				// TODO Auto-generated catch block
				e1.printStackTrace();
			}
		}
		if(player.getDevice() != localDevice) {
			try {
				player.switchDevice(localDevice);
			} catch (IllegalStateException | AudioEngineException e) {
				errorMessage = e.getMessage();
			}
		}

		if(target.isTargetPlaying()) {
			player.start();
		} else {
			player.pause();
		}
	}

    /**
     * Updates the PlaybackStatus to reflect the current status of the audio engine.
     * This may result in the information being sent to connected machines.
     */
	private void publishInfo() {
		boolean playing = player != null && player.isPlaying();
		double position = player != null ? player.getPosition() : 0;
		double duration = player != null ? player.getDuration() : 0;

		info.setStatus(target.getTargetDevice(), supportedTypes, currentFile,
				gain, mute,
				playing, false, errorMessage, position, System.currentTimeMillis(), duration, endOfMediaReached);
	}


	private Optional<AudioDevice> findLocalDevice(Optional<Speaker> speaker) {
		if(!speaker.isPresent()) return Optional.empty();
		String id = speaker.get().getSpeakerId();
		return audio.getDevices().stream().filter(dev -> dev.getID().equals(id)).findAny();
	}

	private boolean isTarget() {
		return findLocalDevice(target.getTargetDevice()).isPresent();
	}


	public Player currentPlayer() {
	    return player;
    }

}

File: src/player/model/playback/PlaybackEngine.kt
package player.model.playback

import audio.AudioDevice
import audio.AudioEngine
import audio.javafx.JavaFXAudioEngine
import audio.javasound.JavaSoundEngine
import cloud.Cloud
import cloud.Peer.Companion.getLocal
import javafx.beans.InvalidationListener
import javafx.beans.binding.Bindings
import javafx.beans.property.SimpleBooleanProperty
import javafx.collections.FXCollections
import javafx.collections.ListChangeListener
import player.model.CycloneConfig
import player.model.data.MasterGain
import player.model.data.PlayTask
import player.model.data.PlayTaskStatus
import player.model.data.Speaker
import java.util.concurrent.Callable
import java.util.concurrent.Executors
import java.util.stream.Collectors

/**
 * Controls the isLocal audio engine.
 * An instance of PlaybackEngine observe a PlayerStatus.
 * Whenever an audio file should be played back on a isLocal device, the PlaybackEngine loads the file and plays it back.
 * It also adjusts position and gain to match the desired status.
 *
 * When a playback event occurs like an error or end-of-file, the PlaybackEngine updates the status.
 */
class PlaybackEngine (val cloud: Cloud, val config: CycloneConfig)
{
    val audioEngine: AudioEngine = createEngine(config.audioEngine.value)
    val mainThread = Executors.newFixedThreadPool(1)
    val jobThreads = Executors.newFixedThreadPool(2)

    private val tasks = cloud.getAll(PlayTask::class.java, this) { r -> mainThread.submit(r) }
    private val masterGainData = cloud.getSynchronized(MasterGain::class.java) { r -> mainThread.submit(r) }

    // Public properties
    val jobs = FXCollections.observableArrayList<Job>()
    val masterGain = Bindings.createDoubleBinding(Callable { masterGainData.value.value }, masterGainData)
    val statusInvalid = SimpleBooleanProperty()

    val speakerMap: Map<Speaker, AudioDevice> = audioEngine.devices.stream().collect(Collectors.toMap({ dev -> Speaker(getLocal(), dev.id, dev.name, dev.minGain, dev.maxGain, dev.isDefault) }, { dev -> dev}))


    init {
//        supportedTypes = ArrayList(audio.supportedMediaTypes.stream().map { t: MediaType -> t.fileExtension }.collect(Collectors.toList()))
        cloud.push(Speaker::class.java, speakerMap.keys, this, true)
        mainThread.submit { tasksUpdated() }
        tasks.addListener(ListChangeListener { mainThread.submit { tasksUpdated() } })
        statusInvalid.addListener(InvalidationListener { if(statusInvalid.value == true) publishInfo() })
    }


    private fun tasksUpdated() {
        // --- Handle existing tasks ---
        for (task in tasks) {
            var job = getJob(task.id)
            if (job == null) {
                if (task.target in speakerMap.keys) {
                    job = getOrCreateJob(task.id)
                }
            }
            job?.task?.value = task
        }
        // --- Handle deleted tasks ---
        for (job in jobs) {
            if (job.task.value !in tasks) {
                job.task.value = null
            }
        }
        // --- Delete jobs ---
        for (job in ArrayList(jobs)) {
            if (!job.isAlive()) {
                if (jobs.none { j -> job in j.references() }) {
                    jobs.remove(job)
                }
            }
        }
    }


    fun getJob(taskId: String): Job? {
        return jobs.firstOrNull { j -> j.taskId == taskId }
    }


    fun getOrCreateJob(taskId: String): Job {
        val existing = jobs.firstOrNull { j -> j.taskId == taskId}
        existing?.let { return existing }
        val newJob = Job(taskId, this, config.bufferTime.value)
        jobs.add(newJob)
        newJob.status.addListener(InvalidationListener { statusInvalid.value = true })
        return newJob
    }


    /**
     * Updates the PlaybackStatus to reflect the current status of the audio engine.
     * This may result in the information being sent to connected machines.
     *
     * Called when [statusInvalid] is set.
     */
    private fun publishInfo() {
        statusInvalid.value = false

        val statuses = ArrayList<PlayTaskStatus>()
        for(job in ArrayList(jobs)) {
            val status = job.status.value
            status?.let { statuses.add(status) }
        }
        cloud.push(PlayTaskStatus::class.java, statuses, this, true)
    }


    fun dispose() {
        jobs.forEach { job -> job.dispose() }
        cloud.yankAll(null, this)
    }

    private fun createEngine(name: String?): AudioEngine {
        if (name == "javafx") {
            return JavaFXAudioEngine()
        }
        if (name == null || name == "java") {
            return JavaSoundEngine()
        }
        throw IllegalArgumentException("Unknown audio engine: $name")
    }

}